python/paddle/fluid/tests/unittests/dygraph_to_static/test_cache_program.py

```diff
@@ -0,0 +1,115 @@
+#   Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+
+import unittest
+import numpy as np
+from collections import Counter
+
+import paddle.fluid as fluid
+
+from paddle.fluid.dygraph.dygraph_to_static import ProgramCache
+from paddle.fluid.dygraph.jit import dygraph_to_static_output
+
+from test_fetch_feed import Pool2D, Linear
+
+
+class TestCacheProgram(unittest.TestCase):
+    def setUp(self):
+        self.batch_num = 5
+        self.dygraph_class = Pool2D
+        self.data = np.random.random((1, 2, 4, 4)).astype('float32')
+
+    def test_cache(self):
+        prev_ops, cur_ops = Counter(), Counter()
+        prev_out, cur_out = None, None
+        main_program = fluid.Program()
+        with fluid.program_guard(main_program):
+            static_net = self.dygraph_class()
+            for batch_id in range(self.batch_num):
+                out = static_net(self.data)
+                # Check outputs
+                prev_out = cur_out
+                cur_out = out
+                # Check forward ops
+                prev_ops = cur_ops
+                cur_ops = Counter([
+                    op.type for op in fluid.default_main_program().block(0).ops
+                ])
+                if batch_id > 0:
+                    self.assertTrue(
+                        np.allclose(prev_out[0], cur_out[0]),
+                        msg='Output in previous batch is {}\n Output in current batch is \n{}'
+                        .format(prev_out, cur_out))
+                    self.assertEqual(prev_ops, cur_ops)
+
+
+class TestCacheProgram2(TestCacheProgram):
+    def setUp(self):
+        self.batch_num = 5
+        self.dygraph_class = Linear
+        self.data = np.random.random((4, 10)).astype('float32')
+
+
+class TestCacheProgramWithOptimizer(unittest.TestCase):
+    def setUp(self):
+        self.dygraph_class = Linear
+        self.data = np.random.random((4, 10)).astype('float32')
+        self.batch_num = 5
+
+    def train_static(self):
+        main_program = fluid.Program()
+        loss_data = []
+        with fluid.program_guard(main_program):
```

Comment:
I can understand that you have to pre-set optimizer and append it to real program only at run time. (because you can only add data feed layer then optimizer) you are doing right thing but API interface name is not natural as we have discussed.

Suggestion:

Make ProgramCache just do cache thing. Use another class to do those whole control thing out of cache (your run logic), like "DygraphToStaticController" or other name if you have a better one, where  your can put running logic (but I am thinking whether now we should give our dygraph to static module a name like AutoGraph for TF, Scripting for PyTorch)

For a Controller class, it can now have method "set_optimizer", much natural.

Reply:
Thanks. I add a new class `AutoTracer` to handle this problem.